sound/soc/sof/pm.c

```diff
@@ -433,12 +433,48 @@ EXPORT_SYMBOL(snd_sof_set_d0_substate);
 
 int snd_sof_resume(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	int ret;
+
+	if (sdev->s0_suspend && sof_ops(sdev)->set_power_state) {
+		/* resuming from D0I3 */
+		dev_dbg(sdev->dev, "ADSP will exit from D0i3...\n");
+		ret = snd_sof_set_d0_substate(sdev, SOF_DSP_D0I0);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to exit from D0I3 %d\n",
+				ret);
+			return ret;
+		}
+
+		/* platform specific resuming from S0ix */
+		return snd_sof_dsp_resume(sdev);
+	}
+
+	/* else, resuming from D3 */
 	return sof_resume(dev, false);
 }
 EXPORT_SYMBOL(snd_sof_resume);
 
 int snd_sof_suspend(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	int ret;
+
+	if (sdev->s0_suspend && sof_ops(sdev)->set_power_state) {
+		/* suspending to S0ix */
```

Comment:
can we please align the comments to something like this? lets not mix up D0i3 and S0ix.
/* DSP exiting out of D0i3 */
/* DSP entering D0i3 */